src/darwin/Framework/CHIP/MTRCallbackBridgeBase.h

```diff
@@ -147,7 +152,23 @@ using MTRActionBlockT = CHIP_ERROR (^)(chip::Messaging::ExchangeManager & exchan
 template <typename SuccessCallback>
 using MTRLocalActionBlockT = CHIP_ERROR (^)(SuccessCallback successCb, MTRErrorCallback failureCb);
 
-template <class T> class MTRCallbackBridge : public MTRCallbackBridgeBase {
+class NoAttestationChallenge {
+};
+
+class HaveAttestationChallenge {
+protected:
+    NSData * mAttestationChallenge;
+};
+
+namespace detail {
+using AttestationResponseCallback
+    = void (*)(void *, const chip::app::Clusters::OperationalCredentials::Commands::AttestationResponse::DecodableType &);
+} // namespace detail
+
+template <class T>
+class MTRCallbackBridge : public MTRCallbackBridgeBase,
```

Comment:
I guess this is fine, but it does seem like a rather complicated way to get some special-case handling into this code.
An alternative would be to simply provide a way to access the session that's being used for an exchange, so the client code can directly grab the attestation challenge from it.

(This class being templated over all response types also makes me worried about code bloat, but maybe the compiler is better at optimizing this away than I give it credit for.)

Reply:
> An alternative would be to simply provide a way to access the session that's being used for an exchange, so the client code can directly grab the attestation challenge from it.

The problem is that by the time the client code is being called (on a different queue) the session may well be gone entirely.  I could just always grab the attestation challenge unconditionally, but that adds more code and a bunch more allocations.

The template is pre-existing, right?  I'm just adding some conditional behavior to it depending on the type of T.  Once it's all compiled it should boil away to nothing for all the types except AttestationResponse....

I would still like to generally make this stuff more generic, obviously.